internal/engine/upper.go

```diff
@@ -25,7 +35,34 @@ func NewUpper(manager service.Manager) (Upper, error) {
 	watcherMaker := func() (watch.Notify, error) {
 		return watch.NewWatcher()
 	}
-	return Upper{b, watcherMaker}, nil
+	sleeper := time.Sleep
+	return Upper{b, watcherMaker, sleeper}, nil
+}
+
+//makes an attempt to read some events from `eventChan` so that multiple file changes that happen at the same time
+//from the user's perspective are grouped together.
+func (u Upper) coalesceEvents(eventChan chan fsnotify.Event) []fsnotify.Event {
+	var events []fsnotify.Event
+	doneWaitingForChanges := false
+	for !doneWaitingForChanges {
+		// keep grabbing changes until we've gone `watchBufferDurationInMs` without seeing a change
```

Comment:
i'm not convinced that this is a good buffering strategy. the loop might never complete if there are continuous changes. 

one of the big problems that came up in live integration was codebases where the user accidentally watched a log directory so that the tool would just hang forever with no output or indication of why it was broken. you should talk to @jazzdan about how he ended up handling this.

Reply:
Yeah we calculated the rate of file events, and if it was too high printed an error and crashed.

Reply:
I've made it so it'll wait a max of 10 seconds, and then just go with whatever it's seen so far.